src/emmy/differential.cljc

```diff
@@ -1331,20 +960,30 @@
 
 ;; Non-differentiable generic operations
 
-(defmethod g/simplify [::differential] [d]
-  (map-coefficients g/simplify d))
-
-(defmethod g/zero? [::differential] [d]
-  (every? (comp g/zero? coefficient) (bare-terms d)))
-
-(defmethod g/one? [::differential] [d] (one? d))
-(defmethod g/identity? [::differential] [d] (identity? d))
-(defmethod g/zero-like [::differential] [_] 0)
-(defmethod g/one-like [::differential] [_] 1)
-(defmethod g/identity-like [::differential] [_] 1)
-(defmethod g/freeze [::differential] [d]
-  (letfn [(freeze-term [term]
-            (make-term (tags term)
-                       (g/freeze (coefficient term))))]
-    `[~'Differential
-      ~@(mapv freeze-term (bare-terms d))]))
+(defbinary v/= (by-primal v/=))
+(defunary g/negative? (by-primal g/negative?))
+(defunary g/infinite? (by-primal g/infinite?))
+
+
+(defunary g/zero?
+  (fn [dx]
+    (let [[p t] (primal-tangent-pair dx)]
+      (and (g/zero? p)
+           (g/zero? t)))))
+
+(defunary g/one? one?)
+(defunary g/identity? identity?)
+
+(defmethod g/zero-like [::dual] [_] 0)
+(defmethod g/one-like [::dual] [_] 1)
+(defmethod g/identity-like [::dual] [_] 1)
+(defmethod g/freeze [::dual] [d]
+  `[~'Dual
```

Comment:
I swore a saw a dict-like implementation of this somewhere

Reply:
I had a `pprint` implementation for tapes that did that:

https://github.com/mentat-collective/emmy/blob/main/src/emmy/tape.cljc#L285-L296

but I didn't use it for `freeze`. Honestly I don't think these should ever make it out into a situation where we'd inspect the results of a freeze...